unix_integration/src/resolver.rs

```diff
@@ -837,31 +900,246 @@ where
+                let auth_session = AuthSession::InProgress {
+                    account_id: account_id.to_string(),
+                    id,
+                    token: token.map(Box::new),
+                    online_at_init,
+                    cred_handler,
+                };
+
+                // Now identify what credentials are needed next. The auth session tells
+                // us this.
+
+                Ok((auth_session, next_req.into()))
+            }
+            Err(IdpError::NotFound) => Ok((AuthSession::Denied, PamAuthResponse::Unknown)),
+            Err(IdpError::ProviderUnauthorised) | Err(IdpError::Transport) => {
+                error!("transport error, moving to offline");
+                // Something went wrong, mark offline.
+                let time = SystemTime::now().add(Duration::from_secs(15));
+                self.set_cachestate(CacheState::OfflineNextCheck(time))
+                    .await;
+                Err(())
+            }
+            Err(IdpError::BadRequest) => Err(()),
+        }
+    }
+
+    pub async fn pam_account_authenticate_step(
+        &self,
+        auth_session: &mut AuthSession,
+        pam_next_req: PamAuthRequest,
+    ) -> Result<PamAuthResponse, ()> {
         let state = self.get_cachestate().await;
-        let (_expired, token) = self.get_cached_usertoken(&id).await?;
 
-        match state {
-            CacheState::Online => self.online_account_authenticate(&token, &id, cred).await,
-            CacheState::OfflineNextCheck(_time) => {
-                // Always attempt to go online to attempt the authentication.
-                if self.test_connection().await {
-                    // Brought ourselves online, lets check.
-                    self.online_account_authenticate(&token, &id, cred).await
+        let maybe_err = match (&mut *auth_session, state) {
+            (
+                &mut AuthSession::InProgress {
+                    ref account_id,
+                    id: _,
+                    token: _,
+                    online_at_init: true,
+                    ref mut cred_handler,
+                },
+                CacheState::Online,
+            ) => {
+                let maybe_cache_action = self
+                    .client
+                    .unix_user_online_auth_step(account_id, cred_handler, pam_next_req)
+                    .await;
+
+                match maybe_cache_action {
+                    Ok((res, AuthCacheAction::None)) => Ok(res),
+                    Ok((
+                        AuthResult::Success { token },
+                        AuthCacheAction::PasswordHashUpdate { cred },
+                    )) => {
+                        // Might need a rework with the tpm code.
+                        self.set_cache_userpassword(token.uuid, &cred).await?;
+                        Ok(AuthResult::Success { token })
+                    }
+                    // I think this state is actually invalid?
```

Comment:
This raises a problem though. How do we cache the password for a user who hasn't completed authentication, and doesn't have a token?

Reply:
If your user needs MFA to authenticate, then you can't cache the pw anyway, because moving from online -> offline would downgrade you to single factor. This would be an MFA bypass I think. The exception here being that MFA is prompted for "once" on first login, and then never again. But if MFA (ie device code) prompts ever time, then it can't proceed.

As well saying this device code already *is* an MFA itself since you need MFA to auth to the browser to be able to validate the device code, so you shouldn't need to pair device code with password anyway? Or am I missing something ....